python/paddle/v2/__init__.py

```diff
@@ -85,22 +79,63 @@ def set_env(key, value):
         if envset is None:
             os.environ[key] = value
 
-    ht = os.popen("lscpu |grep \"per core\"|awk -F':' '{print $2}'|xargs")
-    ht = int(ht.read())
-    if ht == 1:  # ht is off
-        set_env("OMP_DYNAMIC", "false")
-        set_env("KMP_AFFINITY", "granularity=fine,compact,0,0")
-    else:
+    def is_hyperthreading_enabled():
+        '''If Hyper Threading is enabled.'''
+        if platform.system() == "Linux":
+            percore = os.popen(
+                "lscpu |grep \"per core\"|awk -F':' '{print $2}'|xargs")
+            return int(percore.read()) != 1
+        elif platform.system() == "Darwin":
+            physical = int(os.popen("sysctl hw.physicalcpu").read())
+            logical = int(os.popen("sysctl hw.logicalcpu").read())
+            return logical > physical
+        else:
+            # do not support on other platform yet
+            return False
+
+    def get_logical_processors():
+        '''Get the logical processors number'''
+        import platform
+        if platform.system() == "Linux":
+            processors = os.popen(
+                "grep \"processor\" /proc/cpuinfo|sort -u|wc -l")
+            return int(processors.read())
+        elif platform.system() == "Darwin":
+            processors = os.popen("sysctl hw.logicalcpu")
+            return int(processors.read())
+        else:
+            # do not support on other platform yet
+            return 1
+
+    if is_hyperthreading_enabled():
         set_env("OMP_DYNAMIC", "true")
         set_env("KMP_AFFINITY", "granularity=fine,compact,1,0")
-    processors = os.popen("grep \"processor\" /proc/cpuinfo|sort -u|wc -l")
-    processors = int(processors.read())
-    trainers = kwargs.get('trainer_count', 1)
-    threads = processors / trainers
+    else:
+        set_env("OMP_DYNAMIC", "false")
+        set_env("KMP_AFFINITY", "granularity=fine,compact,0,0")
+    processors = get_logical_processors()
+    threads = processors / trainer_count
     threads = '1' if threads < 1 else str(threads)
     set_env("OMP_NUM_THREADS", threads)
     set_env("MKL_NUM_THREADS", threads)
 
+
+def init(**kwargs):
+    import py_paddle.swig_paddle as api
+    args = []
+    args_dict = {}
+    # NOTE: append arguments if they are in ENV
+    for ek, ev in os.environ.iteritems():
+        if ek.startswith("PADDLE_INIT_"):
+            args_dict[ek.replace("PADDLE_INIT_", "").lower()] = str(ev)
+
+    args_dict.update(kwargs)
+    # NOTE: overwrite arguments from ENV if it is in kwargs
+    for key in args_dict.keys():
+        args.append('--%s=%s' % (key, str(args_dict[key])))
+
+    auto_set_cpu_env(kwargs.get('trainer_count', 1))
```

Comment:
All programs, including `auto_set_cpu_env`, were written to do something *automatically*, so it seems that an accurate name could be `set_omp_mkl_env_vars`.

Reply:
OK, thx.